Asio-Server-and-Client/queue.cpp

```diff
@@ -0,0 +1,118 @@
+#include <cmath>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+
+template<typename T, uint64_t SIZE = 4096, uint64_t MAX_SPIN_ON_BUSY = 40000000>
```

Comment:
MAX_SPIN_ON_BUSY is a bad way to count the time producer waits free slot, it can be really different depending on CPU. I recommend to use max_wait_ms and condition_variable signals